app/controllers/users_controller.rb

```diff
@@ -0,0 +1,69 @@
+class UsersController < ApplicationController
+  def index
+    @users = User.all
+  end
+
+  def login_form
+    @user = User.new
+  end
+
+  def show
+    @user = User.find_by(id: params[:id])
+    if @user.nil?
+      flash[:error] = "Could not find user."
+      redirect_to root_path
+    else
+      flash[:success] = "You are logged in."
+    end
+  end
+
+  def current
+    @user = User.find_by(id: session[:user_id])
+    if @user.nil?
+      flash[:error] = "You must be logged in first!"
+      redirect_to root_path
+    end
+    redirect_to current_user_path
+  end
+
+  def login
+    username = params[:user][:username]
+    user = User.find_by(username: username)
+    user = User.create(username: username) if user.nil?
+
+    if user.id
+      session[:user_id] = user.id
+      flash[:alert] = "#{user.username} logged in"
+      redirect_to root_path
+    else
+      flash[:error] = "Unable to log in"
+      redirect_to root_path
+    end
+  end
+
+  def logout
+    user = User.find_by(id: session[:user_id])
+    session[:user_id] = nil
+    flash[:notice] = "Logged out #{user.username}"
+    redirect_to root_path
+  end
+
+  def upvote
```

Comment:
This method is rather long, and nonfunctional on the site, but I suggest adding a method to the `User`model for upvoting.  That can help you reduce the size of your controller method here.